src/environmentd/src/http.rs

```diff
@@ -187,19 +188,140 @@ pub struct InternalHttpConfig {
     pub tracing_handle: TracingHandle,
     pub adapter_client_rx: oneshot::Receiver<mz_adapter::Client>,
     pub active_connection_count: Arc<Mutex<ConnectionCounter>>,
+    pub promote_leader: oneshot::Sender<()>,
+    pub ready_to_promote: oneshot::Receiver<()>,
 }
 
 pub struct InternalHttpServer {
     router: Router,
 }
 
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct LeaderStatusResponse {
+    pub status: LeaderStatus,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub enum LeaderStatus {
+    /// The current leader returns this. It shouldn't need to know that another instance is attempting to update it.
+    IsLeader,
+    /// The new pod should return this status until it is ready to become the leader, or it has determined that it cannot proceed.
+    Initializing,
+    /// Once we receive this status, we can tell it to become the leader and migrate the EIP.
+    ReadyToPromote,
+}
+
+#[derive(Debug)]
+pub enum LeaderState {
+    IsLeader,
+    // Invariant: these options are always Some. The reason for the extra layer is to
+    // allow us to take a &mut LeaderState and access a oneshot::Sender by value.
+    Initializing {
+        promote_leader: Option<oneshot::Sender<()>>,
+        ready_to_promote: Option<oneshot::Receiver<()>>,
+    },
+    // Same invariant as above
+    ReadyToPromote {
+        promote_leader: Option<oneshot::Sender<()>>,
+    },
+}
+
+fn state_to_status(state: &LeaderState) -> LeaderStatus {
+    match state {
+        LeaderState::IsLeader => LeaderStatus::IsLeader,
+        LeaderState::Initializing { .. } => LeaderStatus::Initializing,
+        LeaderState::ReadyToPromote { .. } => LeaderStatus::ReadyToPromote,
+    }
+}
+
+pub async fn handle_leader_status(
+    State(state): State<Arc<Mutex<LeaderState>>>,
+) -> impl IntoResponse {
+    let mut leader_state = state.lock().expect("lock poisoned");
+    match &mut *leader_state {
+        LeaderState::IsLeader => (),
+        LeaderState::Initializing {
+            promote_leader,
+            ready_to_promote,
+        } => {
+            match ready_to_promote.as_mut().expect("invariant").try_recv() {
+                Ok(_) => {
+                    *leader_state = LeaderState::ReadyToPromote {
+                        promote_leader: promote_leader.take(),
```

Comment:
maybe panic if `promote_leader` is `None` here? If it's missing that indicates some logic error